src/elastio-snap.c

```diff
@@ -3420,13 +3420,35 @@ static int snap_trace_bio(struct snap_device *dev, struct bio *bio){
 	struct tracing_params *tp = NULL;
 	sector_t start_sect, end_sect;
 	unsigned int bytes, pages;
+	int max_sectors;
 
 	//if we don't need to cow or physical memory usage has exceeded threshold or
 	//	COW file state is failed, this bio just call the real mrf normally
 	if (!bio_needs_cow(bio, dev) || memory_is_too_low(dev) || tracer_read_cow_fail_state(dev)) {
 		return elastio_snap_call_mrf(dev->sd_orig_mrf, bio);
 	}
 
+	max_sectors = queue_max_sectors(bdev_get_queue(dev->sd_base_dev));
+	if (bio_sectors(bio) > max_sectors) {
+#ifdef HAVE_BIO_SPLIT_4
+		struct bio *split = bio_split(bio, max_sectors,
+				GFP_NOIO, dev_bioset(dev));
+		bio_chain(split, bio);
+#ifdef HAVE_GENERIC_MAKE_REQUEST
+		generic_make_request(bio);
+#else
+		submit_bio_noacct(bio);
+#endif
+		bio = split;
+#else
+		struct bio_pair *bp = bio_split(bio, max_sectors);
```

Comment:
It seems this variable declaration may cause a compilation warning on kernels where `HAVE_GENERIC_MAKE_REQUEST` is false.

Reply:
Why do you think so? It's all under that #else, so no warnings are expected. Just in case, checked it on fedora35 (define not present) and debian9 (define present)

Reply:
> Why do you think so?

Because of the variable declaration not at the beginning of the code block or function. I was mistaken with `HAVE_GENERIC_MAKE_REQUEST`. It's about `HAVE_BIO_SPLIT_4`. I've checked it too. And define is present on Fedora 35 as well as on Debian 9 and 8. It's not present on CentOS 7, but there is no warning in the build log. I was thinking about `warning: ISO C90 forbids mixed declarations and code`.  
But it seems it's not happening even on the oldest supported OSes. Maybe it may happen just on the already unsupported CentOS 6 where GCC uses C89 standard or so )